server/src/test/java/org/elasticsearch/cluster/IndexingWithMappingUpdateIT.java

```diff
@@ -0,0 +1,193 @@
+            .create(createIndexRequest(index)
+                .settings(Settings.builder()
+                    .put("index.number_of_replicas", 1)
+                    .put("index.number_of_shards", 1)))
+            .get());
+
+
+        client = cluster.client(primaryNodeName);
+
+        // create index in advance to have allocate primary
+        client.prepareIndex(index, type, "d").setSource("{ }", XContentType.JSON).get();
+
+        replicaNodeName = cluster.startDataOnlyNode();
+
+        ensureStableCluster(3);
+
+        ClusterState clusterState = client().admin().cluster().prepareState().get().getState();
+        String nodeId = clusterState.getRoutingTable().index(index).shard(0).primaryShard().currentNodeId();
+        assertThat(clusterState.getRoutingNodes().node(nodeId).node().getName(), is(primaryNodeName));
+    }
+
+    /**
+     title Replication Sequence, case 1
+
+     Client -> Primary: doc1
+     Primary -> Master: Update mapping
+     Master -> Primary: mapping update
+     Primary -> Master: ack mapping update
+     Primary -> Primary: updates local mapping
+
+     Client -> Primary: doc2
+     Primary -> Replica: doc2
+
+
+     Master -> Replica: mapping update
+     Replica -> Master: ack mapping update
+     Master -> Primary: all nodes acked
+     Primary -> Replica: doc1
+     */
+    public void testRaiseConditionOnSecondDocWithMappingUpdate() throws Exception {
+        int docs = randomIntBetween(5, 10);
+
+        final CountDownLatch prestartLatch = new CountDownLatch(1 + docs);
+        final CountDownLatch startLatch = new CountDownLatch(1);
+        final CountDownLatch sendSecondDocLatch = new CountDownLatch(1);
+        final CountDownLatch finishLatch = new CountDownLatch(1 + docs);
+
+        AtomicInteger bulkMessagesCounter = new AtomicInteger();
+
+        MockTransportService replicaTransportService =
+            ((MockTransportService) cluster.getInstance(TransportService.class, replicaNodeName));
+
+        // tracer on replica to slow down consumption of some requests :
+        // internal:discovery/zen/publish/sent , internal:discovery/zen/publish/commit - bring cluster state with mapping update
+        // indices:data/write/bulk - indexing request
+        replicaTransportService.addTracer(new MockTransportService.Tracer() {
+            @Override
+            public void receivedRequest(long requestId, String action) {
+                boolean clusterStateCommit = action.equals(PublishClusterStateAction.COMMIT_ACTION_NAME);
+                boolean clusterStateWrite = action.equals(PublishClusterStateAction.SEND_ACTION_NAME);
+                boolean firstBulkRequestSend = bulkMessagesCounter.getAndIncrement() == 0 &&
+                    action.startsWith(BulkAction.NAME);
+
+                if (clusterStateCommit) {
+                    sendSecondDocLatch.countDown();
+                }
+                if (clusterStateCommit || clusterStateWrite || firstBulkRequestSend) {
```

Comment:
AIUI we should only have to slow down the delivery of the commit of the mapping update on the replica to trigger the behaviour we're after. Are these other cases also necessary?